### ISSUE TITLE
New rule suggestion: require-style

### ISSUE THREAD

**Issue by denis-sokolov** [eslint/eslint#2555]
_Combining ideas of #2539, #2541, #2542 into one rule. If this looks good, we shall close the other proposals._

I would like to add a new rule to eslint - `require-style`.
It would allow enforcing a particular module structure by validating local file `require` calls against multiple possible options.

The rule is very specific to a particular workflow, and definitely should be off by default (this choice is irrelevant with 1.0.0 using `--reset`).

# Options

## Extension

Whether to require, forbid, or ignore file extension in `require` calls. Perhaps a configurable list of extensions that need to be left out.

The following would be invalid when the option is enabled:
```javascript
require('./foo.js');
require('./foo/bar.js');
require('../foo.js');
require('foo/bar.js');
```

The following would still be valid when the rule is enabled:
```javascript
require('./foo');
require('./foo/bar');
require('../foo');
require('foo/bar');
```

The use case for requiring the extension is - there is a common programming pattern where you start with small and simple implementations and you can later grow them as needed. What started as `foo.js` can grow too long and need its own subdirectory to be split off into many files. This transition is smooth if everybody who depends on `foo` only used `require('./foo')`, but if dependents used `require('./foo.js')`, then either dependents need to be modified, or a stub `foo.js` needs to remain.

The error thus would prevent a maintainability error by a developer.

## Up

Whether to forbid requiring up (`../`) entirely, or more than some amount of levels up.

The following would be invalid when the rule is enabled:
```javascript
require('../foo');
require('./../bar');
require('../../bar');
```

The use case is - it is sometimes customary to mirror the project module dependency tree using the filesystem. That makes the project structure considerably easier to understand, it almost guarantees a lack of an intricate web of dependencies inside a project.

The option thus would prevent a semantic error by a developer. It would also prevent developers from the temptation of using the up selector "just this once".

## Down

Whether to forbid requiring into subdirectories (`./foo/bar`), or more than some amount of levels deep.

The following would be invalid when the option is enabled:
```javascript
require('./foo/bar'); // 1 level deep
require('./foo/bar.js');
require('./foo/bar/quux'); // 2 levels deep
require('../foo/bar');
```

If the allowed number of levels is, say, 1, only the following among the previous would be invalid:
```javascript
require('./foo/bar/quux'); // 2 levels deep
```

The following would still be valid when the option is enabled:
```javascript
require('./foo'); // Current (0-th) level
require('../foo'); // Up
```

The option intent, module structure, has no effect on 3rd party modules, thus this would be valid regardless of the option setting:
```javascript
 require('unicode-7.0.0/categories/Lu/symbols');
```

**Comment by ilyavolodin** [comment]
I think this rule can be useful, I'm not sold on the name. To me `require-style` is something that will deal with style of require statement, white spaces around it and such.

**Comment by mathieumg** [comment]
As stated in the aforementioned issues, I believe there is room for this rule. However, I'm not convinced about the name either. Would it also apply to ES6 `import` statements? I believe it would need to. (And thus a better name would be welcome!)

**Comment by denis-sokolov** [comment]
Perhaps `relative-import-style`?

**Comment by ilyavolodin** [comment]
I feel like there are two keywords for this rule: `import` and `path`, since this will not affect npm module imports.

**Comment by denis-sokolov** [comment]
That's why I thought `relative` would be a good keyword. But `import-path-style` sounds just as well.

**Comment by nzakas** [comment]
How about `require-restrict` or `restrict-require`?

**Comment by mathieumg** [comment]
Perhaps `restrict-module` or even `restrict-module-path`? If we want to cover both `require` and `import`.

**Comment by nzakas** [comment]
I don't know that we want to cover both with one rule

**Comment by ilyavolodin** [comment]
I would vote for `restrict-require` or `restrict-require-path` and make it apply to `require` statements only for now. If we get requests down the line for `import` support, we can add it, and rename the rule before 2.0.0.

**Comment by mathieumg** [comment]
> I don't know that we want to cover both with one rule

Why not? It only applies to the "path" part for both, so the distinction wouldn't serve any purpose:

```js
require("PATH");
import a from "PATH";
import {a} from "PATH";
import a, {b} from "PATH";
import a, {b as c} from "PATH";
// etc.
```

**Comment by denis-sokolov** [comment]
Frankly, indeed, `require` and `import` has identical semantics.
The intent of the rule is to restrict which modules are imported into the scope.
The syntax used for that does not matter.

**Comment by xjamundx** [comment]
Would someone need to enable `ecmaFeatures: { module: true }` to use this feature for `require` for example?

**Comment by denis-sokolov** [comment]
No, that requirement would not be beneficial. What about developers that rely on `require` alone and the current Node platform?

**Comment by nzakas** [comment]
@denis-sokolov unfortunately, the semantics are very different even though the syntax looks functionally similar.

**Comment by denis-sokolov** [comment]
@nzakas, how so? For our purposes, restricting how deep we can navigate for dependencies, `import foo from 'foo'` is equivalent to `var foo = require('foo')` and `import { bar } from 'foo'` is equivalent to `var bar = require('foo').bar, is that not true?

**Comment by nzakas** [comment]
The semantics of import statements are very different than `require` - while they appear to do the same thing, they actually do not. ES6 modules have their own scoping rules, create lexical bindings via import, etc.; `require` is just a method call and has no scoping or binding effects. So the code you mention is not actually equivalent.

You're talking about there being a logical equivalence between the two, not a semantic equivalence.

My concern here is that we seem to keep expanding the scope of the rule. Can we assume that people will want to use the same restrictions in both places? What about if they want separate restrictions for require vs. import?

**Comment by denis-sokolov** [comment]
Note that I said "for our purposes". For our purposes, that is, restricting relative imports to a particular module structure as defined by the configuration, `var foo = require('foo')` is equivalent to `import foo from 'foo'`. If I want to use the rule and make sure me and my team do not import grand-children, I certainly do not care whether CommonJS or ES6 syntax is used to do so!

In my mind that's the meaning of the rule - restrict relative import to enforce a particular dependency structure in a module. This meaning and this use case has no reason to distinguish between `require` and `import`. Can you provide a use case and a meaning for the rule that would want to distinguish it?
If you want to stay safe, we can, of course, make two separate but identical rules, with one only caring about `require`, and another about `import`, I don't mind.
But are you sure we don't need separate rules for `import foo from 'foo'` and `import { foo } from 'foo'`, then?

**Comment by nzakas** [comment]
Yes, logical equivalency, not semantic equivalency - this is an important distinction for tools like ESLint. 

You have to realize, I've been doing this for two years, and so I know the problems that can happen when creating rules that are ill-defined or overly broad. Once rules are introduced, they will be there for a long time, so I have a vested interest in making sure the ones that are bundled with ESLint make sense. I'm also acutely aware that I cannot predict the future, so I've stopped placing bets on what people might or might not do or want.

Nowhere else in ESLint do we treat `require` and `import` the same, and this wasn't even part of your original proposal, so it naturally is setting off warning bells for me. 

Of course, you could also decide to write whatever rule you want and publish it as a plugin. That is why we have plugins - we don't want to be the gatekeepers to all the rules in the world. However, bundled rules need to make sense as being part of a coherent whole, so we need to keep the bar for bundled rules quite high.

All that is to say, I don't know that we can assume people want to have the same settings for `require` and `import`, and making that assumption could hurt us in the future. They could want very different settings for each, and how would we support that?

**Comment by denis-sokolov** [comment]
The current proposal looks roughly like this: `import-path-style: { up: 2, down: 1 }`.
If a use case like this arises, we can support it by adding more options to the rule:
`import-path-style: { require-up: 3, import-up: 1, down: 1 }`, while allowing the common use case to use a shorthand of `up` and `down` options.
If you dislike that, should we make two rules `import-path-style` and `require-path-style`?

**Comment by gyandeeps** [comment]
@eslint/eslint-team Suggestions or feedback?

**Comment by IanVS** [comment]
From the original description:

> The rule is very specific to a particular workflow

Therefore I agree with the suggestion that this be written as a plugin.

**Comment by nzakas** [comment]
I agree. I think in cases where we can't find a clear resolution, the best path forward is to suggest a custom rule/plugin. If and when such a rule/plugin is proven out, we can always decide it's important enough to include the core.
